Feature/http cache (not to merge)

Sivaramvt opened the pull request:
HttpCache. 60-70% implementation complete.
Created stub for Disk and JCR based cache store. Planning to involve @jasiwalgaurav for implementing these.

@davidjgonzalez Created a pull request against my fork master to get your opinion on the code base. Your idea of this http cache is really nice. We would want to make this really usable in projects. Please share your review comments.

BTW, I made this a separate bundle as there was a OSGi package dependency conflict and bundle was not resolving. Once implementation is complete, I will take your guidance on the packaging.

Sivaramvt committed:
Detailed design added for HttpCacheEngine and HttpCacheConfig. HttpCacheConfig modelled as config factory.

Sivaramvt committed:
Added detailed design for cache handling rules and cache data store.

Sivaramvt committed:
Added detailed design for sling filter for http cache layer.

Sivaramvt committed:
Added detailed design for sling filter for http cache layer.

Sivaramvt committed:
Added business logic methods for HttpCacheEngine.

Sivaramvt committed:
Added classes and implementation for cache invalidation package.

Sivaramvt committed:
Added implementation for the filter and configs.

Sivaramvt committed:
Added CacheKey and CacheContent. Modified pom to use servlet 3.0 APIs.

Sivaramvt committed:
Adding APIs for CacheStore.

Sivaramvt committed:
Added Google Guava v 15.0 in pom. Guava is required for the men cache store implementation.

Sivaramvt committed:
Implemented MemCacheStore implementation.

Sivaramvt committed:
Added Mbean for MemCacheStore.

Sivaramvt committed:
Added place holder for JCR Cache Store.

Sivaramvt committed:
Moving http cache issue to a new bundle due to java.inject dependency conflict.

Sivaramvt committed:
Moving the files to httpcache bundle.

Sivaramvt committed:
Removing the changes made in main bundle project.

Sivaramvt committed:
Tweaking the bundle pom to use servlet 3 APIs.

Sivaramvt committed:
Changes in APIs of Cache Config.

Sivaramvt committed:
Stuffing skeleton implementation for major methods.

Sivaramvt committed:
Added few rules.

Sivaramvt committed:
Adding javadoc.

davidjgonzalez commented on bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/config/impl/HttpCacheConfigImpl.java:
```diff
@@ -0,0 +1,131 @@
+public class HttpCacheConfigImpl implements HttpCacheConfig {
+    private static final Logger log = LoggerFactory.getLogger(HttpCacheConfigImpl.class);
+
+    @Property(label = "Request URI pattern",
```
Any reason to not make this multiple patterns? I imagine a  common use cases would look something like

`Whitelist: [ /content/mysite/(.*).html,  /content/mysite/(.*).json ]`

Sivaramvt replied:
2 reasons for this multiple configs.
(1) Ability to state where this cached response for the request should be persisted. For example, I could say all zip (._).zip to be persisted in disk and all Json  (._).json to be in-mem at the same time.
(2) Ability to invalidate the cache selectively. Example - I could put up a config saying all request for product data /content/mysite(.*).product-data.json to be invalidated for any changes in /etc/my-site/my-product-data without disturbing other cached items.

davidjgonzalez commented on bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/config/impl/HttpCacheConfigImpl.java:
```diff
@@ -0,0 +1,131 @@
+                      ".json. Mandatory parameter.")
+    private static final String PROP_REQUEST_URI_PATTERN = "httpcache.config.requesturi.pattern";
+    private String requestUriPattern;
+
```
It would be useful to have a Blacklist that is evaluated after the whitelist (Whitelist being "Request URI Pattern") .. This way it would be very easy to include `*.html` but exclude `*.nocache.*`

Sivaramvt replied:
We could hook in custom rules for providing such exclusions. Rules implementation is very powerful idea. Lifted from Granite Http Cache impl :)

Reference
https://github.com/Sivaramvt/acs-aem-commons/blob/feature/http-cache/bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/rule/HttpCacheHandlingRule.java

Sample implementation
https://github.com/Sivaramvt/acs-aem-commons/blob/feature/http-cache/bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/rule/impl/CacheOnlyGetRequest.java

Please let me know if that sounds fine.

davidjgonzalez commented on bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/config/impl/HttpCacheConfigImpl.java:
```diff
@@ -0,0 +1,131 @@
+    private static final String PROP_REQUEST_URI_PATTERN = "httpcache.config.requesturi.pattern";
+    private String requestUriPattern;
+
+    @Property(label = "Is request authenticated",
```
WDYT about having this be 3 options:

1) Only anonymous requests
2) Only authenticated requests
3) All requests

Sivaramvt replied:
Noted. Add to TODO list. Let me see how best to factor in these 3 options in a single config.

davidjgonzalez commented on bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/config/impl/HttpCacheConfigImpl.java:
```diff
@@ -0,0 +1,131 @@
+    private static final boolean DEFAULT_IS_REQUEST_AUTHENTICATION_REQUIRED = false;
+    private boolean isRequestAuthenticationRequired;
+
+    @Property(label = "AEM user groups",
```
Am thinking it might be better to have the user define a "CacheKey" ... That way custom cache key implementations should be created.

The default CacheKey provided impl could approximately be the CacheKey in the this PR.

Sivaramvt replied:
Could you please help with little more details. Are we saying, we should provide developers the ability to provide custom implementation for cache key?

davidjgonzalez commented on bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/engine/CacheContent.java:
```diff
@@ -0,0 +1,115 @@
+    /** Response headers */
+    private Map<String, List<String>> headers = new HashMap<>();
+    /** Response content in file */
+    private InputStream dataInputStream;
```
WDYT about capturing the size of the content? Generally thinking it would be interesting to know how big the total cache is.

Sivaramvt replied:
Noted. Will add.

davidjgonzalez commented on bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/engine/TeeServletOutputStream.java:
```diff
@@ -0,0 +1,43 @@
+/**
+ *
+ */
+public class TeeServletOutputStream extends ServletOutputStream {
```
What does "Tee" mean/stand for? :)

Sivaramvt replied:
Originated from https://commons.apache.org/proper/commons-io/apidocs/org/apache/commons/io/output/TeeOutputStream.html :)

davidjgonzalez replied:
very cool - had not seen that before!

davidjgonzalez commented on bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/config/impl/HttpCacheConfigImpl.java:
```diff
@@ -0,0 +1,131 @@
+
+        cacheStore = PropertiesUtil.toString(configs.get(PROP_CACHE_STORE), StringUtils.EMPTY).trim();
+
+        cacheInvalidationPathPattern = PropertiesUtil.toStringArray(PROP_CACHE_INVALIDATION_PATH_PATTERN, ArrayUtils
```
WDYT about compiling these into Patterns in activate; Compiling patterns is generally the expensive part of regex.

Sivaramvt replied:
We should. Will handle it.

davidjgonzalez commented on bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/invalidator/event/JCRNodeChangeEventHandler.java:
```diff
@@ -0,0 +1,63 @@
+                              value = "(|(" + SlingConstants.PROPERTY_PATH + "=" +
+                                      "/content*)(" + SlingConstants.PROPERTY_PATH + "=" + "/etc*))",
+                              description = "Event filter announcing that this is " +
+                                      "interested only for the MPAS " + "products parent " +
```
Description should be use agnostic wording :)

Sivaramvt replied:
Copy + Paste issue :) Will correct it.

davidjgonzalez commented on bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/rule/impl/CacheOnlyGetRequest.java:
```diff
@@ -0,0 +1,23 @@
+    @Override
+    public boolean onRequestReceive(SlingHttpServletRequest request) {
+        // Return true only if Http method is GET.
+        if (!request.getMethod().equals(HTTP_GET_METHOD))
```
missing { .. }  or just `return request.getMethod().equals(HTTP_GET_METHOD)`

Sivaramvt replied:
Noted. Corrected.

davidjgonzalez commented on bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/store/HttpCacheStore.java:
```diff
@@ -0,0 +1,63 @@
+     *
+     * @return
+     */
+    long size();
```
Ah - here is the size .. gotcha

davidjgonzalez commented:
@Sivaramvt generally this looks really good. Added some comments around configuration aspects. The general mechanisms look quite good. Depending on timing, it may make sense to just impl the InMem cache; more timing than anything else on those.

Sivaramvt commented:
Thanks @davidjgonzalez for the inputs. Will work on them and come back to you with a working In-Mem http cache. :)

Sivaramvt committed:
Add TODOs for addressing review comments.

Sivaramvt committed:
Revamped cache config addressing review comments.

Sivaramvt committed:
Implemented isRequestCacheable and getCacheConfig vertical slice.

davidjgonzalez committed:
Updates to support CacheKeyFactory; Removed copy/paste tests

Sivaramvt committed:
Merge pull request #2 from davidjgonzalez/feature/http-cache

Feature/http cache

Sivaramvt committed:
Merge branch 'master' of https://github.com/Sivaramvt/acs-aem-commons into feature/http-cache

# By David Gonzalez
# Via David Gonzalez (1) and Sivaramvt (1)
* 'master' of https://github.com/Sivaramvt/acs-aem-commons:
  Updates to support CacheKeyFactory; Removed copy/paste tests

Conflicts:
	bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/engine/impl/HttpCacheEngineImpl.java

Sivaramvt committed:
Implemented the isRequestCacheable use-case vertically.

Sivaramvt committed:
Minor changes required to get the isRequestCacheable + getCacheConfig methods working.

Sivaramvt committed:
Changes to get the in-mem implementation working.

Sivaramvt committed:
Cleansing the codebase.

Sivaramvt commented:
@davidjgonzalez First attempt, tried with a 26MB zip file having assets. It was taking 2.5s to download. After caching it in mem, it was taking just 0.05s. Wondering how much expensive repo reads and disk access are.

davidjgonzalez commented:
@Sivaramvt measuring cache perf against raw data (ex. 26MB zip) can be a bit tricky since AEM and (many/most) OS's have built in caching. Ex. TarMK uses off-heap memory to cache segment data (if you have a repo that is 10GB and you have 10GB+ of free off-heap mem, the entire repo should be moved to memory.

Might be interesting to look at /system/console/requests and see if there is any part of the the usual request trace that is taking alot of time you're skipping.

I'd think this cache will really shine for computationally heavy pages (ex. query driven or computational complex pages, etc.) -- Pretty cool that it also really helps with "dumb" reads too though!!!

Sivaramvt commented:
@davidjgonzalez Oh. Got it.

> "Now imagine if a different customer wanted to separate their U1 is cached, but instead of Groups, they use Profile Attributes (P1, P2) .. so the cache key would be: U1 + P1, U1 + P2"

Agree with you. Above requirement is very much valid and we need to address it.

However, I still have confusion around the proposed implementation. Please help me get it right. Making this little verbose, so that we get on to the same page.

IMO,
CacheKey is 
- internal to the program
- mapped one on one with cached requests
- generated using request + cache config.

Cache config is
- externally visible as configuration.
- mapped to request patterns (not single request).
- Generates multiple cache key.

So, the requirement is stated below. Please confirm if it's right.
- Provide developers the ability to extend cache config. Example - Group based, Profile attribute based, etc.

> Idea 1: Define on CacheKeyFactory an accepts(..) method, that is called in HttpCacheConfigImpl.accepts(..) (where the current user group check exists).

Does it not tying user groups to CacheKeyFactory looks like spilling the configuration details in multiple places.  Also, how can we make this configuration change with cache config. May be, am missing your line of thought. Please help me to get it right.

This subject is tricky. I was thinking about it. How about making HttpCacheConfigImpl abstract osgi component and force the implementations to override methods such as accept. While extending abstract HttpCacgeConfigImpl we get the ability to add additinal OSGi configs. That means that, developers have to supply custom cache config and correspoding cache key factory implementations. We could supply GroupBasedHttpCacheConfigImpl and GroupCacheKeyFactory. Please let me know what you think.

davidjgonzalez commented:
@Sivaramvt agree on all points. Your last paragraph gets to the root of the issue..

The current HttpCacheConfigImpl contains 95% common code we wouldn't want ACS Devs to re-create. I like the idea of AbstractHttpCacheConfig, w the reservation of abstract OSGi @Components are deprecated -- which is why i was trying to figure out how to separate the piece most ppl would want to extend (which conceptually the key of the cache; the combo of aspects associated w the requests ).

If abstract @OSGi Components werent deprecated, i think its the perfect approach.

http://javadox.com/org.apache.felix/org.apache.felix.scr.annotations/1.9.8/org/apache/felix/scr/annotations/Component.html

Sivaramvt committed:
Major Changes (1) Exposed developer hooks for extension in cache config. (2) Implemented JMX for the men cache store.

Sivaramvt commented:
@davidjgonzalez Glad that we gained the shared understanding. 
I've provided a custom hook (HttpCacheConfigExtension) for developers  to extend the HttpCacheConfig. In the HttpCacheConfigImpl, I exposed a custom attribute which is a string array that will be supplied to the extension to ensure that all cache configurations stay in the same UI.

Please pull the latest code and see how it looks.

Cache config felix config -- <img width="1509" alt="cache config" src="https://cloud.githubusercontent.com/assets/7424268/11372387/3050a2f6-92f3-11e5-82b1-c7b21c2fb315.png">

davidjgonzalez committed:
Combined Extension and CacheKeyFactory; Allows for this custom Group-based implementation to define its own OSGi property set. Allows for removal of the HttpCacheConfigImpl's customAttributes which may or may not be sufficient for future use-cases.

davidjgonzalez committed:
#typo

davidjgonzalez committed:
Enabled/disable OSGi Component by making its dependencies mandatory; Defining at HttpCacheConfigImpl config will be sufficient to enable the Engine

davidjgonzalez committed:
Filter order and timing

davidjgonzalez committed:
New MBean method to view contents of cache; Minor fixes to other MBean displays.

davidjgonzalez committed:
Resolved exception where MultiMaps wrappedSet could not be converted to a List<String> for the header values.

davidjgonzalez committed:
Removed GroupKeyFactory as this is now part of GroupHttpConfigExtension

davidjgonzalez committed:
Enhancements to CacheStats Mbean

Sivaramvt committed:
Merge pull request #3 from davidjgonzalez/feature/http-cache

Feature/http cache

Sivaramvt commented:
Just wanted to record this again for reference in this pull request.

@davidjgonzalez I made this config factory and move along with cacheconfig (n - n mapping). Result is so beautiful. For the same uri, i was able to cache separately for 2 different groups.

<img width="1512" alt="cache-config-geometrix" src="https://cloud.githubusercontent.com/assets/7424268/11422636/a4d16cd4-9463-11e5-8d74-dc7dfb638d70.png">
<img width="1514" alt="cache-config-admin" src="https://cloud.githubusercontent.com/assets/7424268/11422637/a4d38316-9463-11e5-9374-039906a2242c.png">

sling:OSGi config nodes for GroupHttpCacheConfigExtension
com.adobe.acs.commons.httpcache.config.impl.GroupHttpCacheConfigExtension-admins.xml
<?xml version="1.0" encoding="UTF-8"?>
<jcr:root xmlns:sling="http://sling.apache.org/jcr/sling/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0"
          jcr:primaryType="sling:OsgiConfig"
          config.name="group-admins"
          httpcache.config.extension.user-groups="administrators"/>

com.adobe.acs.commons.httpcache.config.impl.GroupHttpCacheConfigExtension-geo.xml
<?xml version="1.0" encoding="UTF-8"?>
<jcr:root xmlns:sling="http://sling.apache.org/jcr/sling/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0"
          jcr:primaryType="sling:OsgiConfig"
          config.name="group-geo"
          httpcache.config.extension.user-groups="geometrixx-twitter"/>

Sivaramvt committed:
Converted GroupHttpCacheConfigExtension into a osgi config factory so that it moves along 1-1 with HttpCacheConfig.

Sivaramvt commented:
MBeans implementation is so handy to have complete control over the cache. Making me glued to it :)

Sivaramvt committed:
Minor tweaks in documentation on GroupHttpCacheConfigExtension.

Sivaramvt committed:
Renamed an API in HttpCacheConfig.

davidjgonzalez committed:
MBean value for Load Exception Rate to %

davidjgonzalez committed:
.equals off a value we control as being non-null

davidjgonzalez committed:
missing brackets for readability

davidjgonzalez committed:
Rule OSGi annotations
- Added missing @Service annotations
- Added metataype=true on disabled @Components to allow for easy activation from Felix
- typos

davidjgonzalez committed:
Updated MBean to report cache size in human-friendly format

davidjgonzalez committed:
Cleaned up rule OSGi annotations
- Removed the metatype true from DoNoCacheRequestWithQueryString and MarkResponseAsCacheDelivered as saving from configMgr does not create the expected empty config

davidjgonzalez committed:
Updated @OSGi property name to make it more idiomatic and reduce collisions with future Extensions

davidjgonzalez committed:
Support of HttpCacheConfig order priority
- Added ability to the order of HttpCacheConfigs to be specified via OSGi @Property, and the the first accepting HttpCacheConfig handles the request.

davidjgonzalez committed:
Cleaned up MBean and added extra info to the Cache Entries

davidjgonzalez committed:
HttpCacheEngine MBean - leverage cacheConfig method rather than OSGi config map

Sivaramvt commented:
@davidjgonzalez Liked all the changes. We would have to deal with the random config selection when there is tie in ordering. I think that's potentially risky.

davidjgonzalez committed:
Support HttpCacheConfigConflicts on configurations that have the same order priority and accept the same request.
If 2 configs accept the same request, the order must be different.

davidjgonzalez commented:
@Sivaramvt click on Pull Requests tab at the top, click on the PR, which will take you here!

Sivaramvt commented:
@davidjgonzalez Am trying to merge the below change. Weird, I don't see that listed anywhere including pull requests. May be, I should take a crash course on GH :)
https://github.com/Sivaramvt/acs-aem-commons/commit/14bc3aa9aeebda02a0438d5aa35937d0b93ae9eb#commitcomment-14664781

Sivaramvt commented on bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/engine/impl/HttpCacheEngineImpl.java:
```diff
@@ -1,77 +1,108 @@
 
     /** Thread safe list to contain the registered HttpCacheConfig references. */
-    private static CopyOnWriteArrayList<HttpCacheConfig> cacheConfigs = new CopyOnWriteArrayList<>();
+    private static final ConcurrentSkipListSet<HttpCacheConfig> cacheConfigs = new ConcurrentSkipListSet<>(new HttpCacheConfigComparator());
```
@davidjgonzalez There is an inconsistency as we use ConcurrentSkipListSet for tracking cache config. ConcurrentSkipListSet uses the given Comparitor for comparing and checking uniqueness. With this current implementation, we would be able to have only one element for the given cache config order at any time. 

We may have to go back to CopyOnWriteArrayList/Set and take care of ordering in the HttpCacheConfig.getCacheConfig method. What do you think?

davidjgonzalez replied:
@Sivaramvt yup - looks right. WDYT about synchronizing the object and managing the sort order in the bind method? Bind should be an extremely rare operation (in comparison to the # of requests that will go through this stack) so dont expect we would be hurt by locks.

Sivaramvt committed:
Rules - Corrected the array representation in HonorCacheControlHeaders.

davidjgonzalez commented:
@Sivaramvt and btw - that last set i pushed direct to your repo/branch it seems; i didnt update the dest repo in gitbox before i pushed (i has pulled off yours and didnt update it back) - so i think youve gotta figured out :)

davidjgonzalez committed:
Handle multiple HttpCacheConfigs w the same order

davidjgonzalez committed:
Minor efficiency in HttpCacheConfig checks

Sivaramvt committed:
Merge pull request #5 from davidjgonzalez/feature/http-cache

Feature/http cache

davidjgonzalez committed:
Split HttpCacheHandlingRules to be configurable at the global Engine level and at the HttpCacheConfig via LDAP filter expressions

davidjgonzalez committed:
Merge branch 'master' of github.com:Adobe-Consulting-Services/acs-aem-commons into feature/http-cache

davidjgonzalez committed:
Introduced TeePrintWriter to address issue of JSON responses not being properly cached.

Sivaramvt committed:
Merge pull request #6 from davidjgonzalez/feature/http-cache

Enhancement and Fix to HTTP Cache

Sivaramvt commented:
@davidjgonzalez Now, with the latest commit, we got rid of (or minimized drastically) the performance penalty attached to the response cache event. Please see if it looks good.

davidjgonzalez commented:
@Sivaramvt if you have time, can you use the approach in the last comment (https://github.com/Adobe-Consulting-Services/acs-aem-commons/issues/594) to pull httpcache into the main acs commons bundle.

If you want, it might be good to git squash all these commits into one big HTTP Cache commit for this merge. Youll need to update the acs pom's to use javax.servlet 3.0.1; when you do that about 8 tests will break. Please just delete them from your branch to get it compiling. I am working on updating those tests to work w servlet 3 on my other computer and we can merge them back in when im done.

Would be great if you can get the files copied over mostly so the git history correctly reflects that this work is yours! I just wouldnt feel right if all of a sudden all this came over under my name because of a bulk copy/paste! :)

Sivaramvt commented:
You are too honest. Pleasure working with you @davidjgonzalez . Yep, I would love to have a commit in my name :) 
Not to forget that this is your idea and there is bulk of contribution from you on code. If I do squash, i fear all your commits will go. At the same time, we may not want to flood the origin with unnecessary commit history. Let me experiment.

Will give this a shot today and keep this thread posted.

Sivaramvt committed:
Http Cache - Preparing for merge to origin master.

davidjgonzalez commented:
@Sivaramvt no worries on the squash - you 100% deserve the attribution! im just happy to have this awesome tool as something i can use.

Sounds good - once we get that, ill merge in these test updates and well be ready to roll!! A week late ... thats not _too_ bad ;)

Sivaramvt commented:
@davidjgonzalez All changes done. POM entry, Pulling code in to the main bundle, updating javax.servlet dependency.

I could squash only 15 latest commits as there was a pull and merge from master. When I tried to squash all 78 commits, the diff got screwed up. 

BTW, Shall I create a pull request against origin master? That should get us a good diff view.

Sivaramvt commented:
BTW, the tests didn't stop my build as I skipped them through Maven switch.

davidjgonzalez commented:
@Sivaramvt really? I was getting compile time errors even w tests disabled. I think if you run a clean build, it will attempt to recompile the tests at which point theyll fail the build. 

But - if it works, then thats fine. I can merge over them either way.

davidjgonzalez commented:
@Sivaramvt and yeh - lets PR against master; I'll make a PR against your squashed branch to get the updated/removed tests in. I think for 2 of the tests i'll end up removing for this release since theyre gonna be a serious PITA to move over to new sling mocks; theyre pretty brittle anyhow but ill work on getting them back in.

Sivaramvt commented:
@davidjgonzalez Here's the pull request. Looks like there is a merge conflict but GH not showing me the conflict anywhere. BTW, I've added you to my fork and you should be able to commit the unit test changes directly. 
https://github.com/Adobe-Consulting-Services/acs-aem-commons/pull/600

First cut of doc available here. Will improve it tomorrow.
https://github.com/Sivaramvt/adobe-consulting-services.github.io/blob/http-cache/_posts/acs-aem-commons/features/2015-12-05-http-cache.md

(Time to bed here :) Great day.)

davidjgonzalez committed:
javax.servlet-api 3.0.1 support

davidjgonzalez committed:
Merging origin/master

davidjgonzalez committed:
Merge branch 'javax-servlet-3.0.1-support' of github.com:davidjgonzalez/acs-aem-commons into feature/http-cache

davidjgonzalez committed:
Felix name hint @Properties and 1.6 compiler support

davidjgonzalez committed:
Http Cache - missing package-info.java files

Sivaramvt commented:
Thanks @davidjgonzalez . Closing this as our first mission accomplished. :)

davidjgonzalez commented:
@Sivaramvt thank _you_! I am excited about the possibilities with this one. Was a large undertaking but expect will be very useful for lots of folks.
